Hint Overhaul

KiernanBrown opened the pull request:
- Hints are now spread evenly across all 13 reports rather than being set to 4 per report.
  - This change required a change to the UE4DataTableInterpreter to allow for an int for the number of hints per report passed into Mobile.Process. https://github.com/WaterKH/UE4DataTableInterpreter/pull/8
- Added a SearchComponent that can be used to build a list of ImportantChecks which should be hinted.
- Changed "Vague: By World” hint type to "Vague: By Category" and changed how hints are generated for it. 
  - All important checks are hinted. If these checks are in less than 13 categories, remaining reports would be filled with hints for empty categories (prioritizing categories that are randomized). If these checks are in more than 13 categories, reports will start to contain more than 1 hint.
  - If there is a check “on Sora’s Initial Setup” or “in the Moogle Synthesis Shop” then the reports these hints are also hint another category or an empty category. Each report should have value and these hints are less valuable than others due to ease of checking them.
- Reports no longer hint themselves (Verbose/Vague: By Check) or hint the category that they are found in (Vague: By Category).
- Edited some hints to be more descriptive thanks to changes made by Sonicshadowsilver2.
- Verbose hints now tell you the gummiphone chest number and the area the chest is found in instead of the chest numbers the game uses for reference, thanks to the chest list from Clara_The_Classy.
- Verbose hints now tell you the level up route names instead of Type A/B/C.

Commit:
Hint Overhaul

Hints are now spread evenly across all 13 reports rather than being set to 4 per report.
Added a SearchComponent that can be used to build a list of ImportantChecks which should be hinted.
Added a “Category” hint type where a report will give the number of important checks for a world/category.
Reports no longer hint themselves.
Edited some hints to be more descriptive thanks to changes made by Sonicshadowsilver2.
Verbose hints now tell you the gummiphone chest number and the area the chest is found in instead of the chest numbers the game uses for reference, thanks to the chest list from Clara_The_Classy.
Verbose hints now tell you the level up route names instead of Type A/B/C.

Commit:
Hint Overhaul

Hints are now spread evenly across all 13 reports rather than being set to 4 per report.
Added a SearchComponent that can be used to build a list of ImportantChecks which should be hinted.
Changed "Vague: By World” hint type to "Vague: By Category" and changed how hints are generated for it. A report will give the number of important checks for a world/category.
Reports no longer hint themselves.
Edited some hints to be more descriptive thanks to changes made by Sonicshadowsilver2.
Verbose hints now tell you the gummiphone chest number and the area the chest is found in instead of the chest numbers the game uses for reference, thanks to the chest list from Clara_The_Classy.
Verbose hints now tell you the level up route names instead of Type A/B/C.

Commit:
Update to version 1.2.3

Updated the UE4DataTableInterpreter.
Fixed an issue with setting hint type to none.

Commit:
Merge branch 'main' into swift_hint_overhaul

WaterKH commented on KH3Randomizer/Data/RandomizerService.cs:
```diff
@@ -244,13 +244,17 @@ public class RandomizerService
             }
         }
 
+        public static Dictionary<DataTableEnum, Dictionary<string, Dictionary<string, string>>> GetDefaultOptions()
```
This is a bit of a red flag for me - We should be not opening multiple of the same file to get the default options like line 332 and line 811

KiernanBrown replied:
It looks like those StreamReader calls on line 332 and line 811 were extra and unnecessary and I think I forgot to remove them when I switched over to calling this method. Doesn't look like those StreamReaders were being used for anything else with how things are currently, so I removed them and this should fix this

WaterKH commented on KH3Randomizer/Pages/Index.razor:
```diff
@@ -88,37 +88,17 @@
+        "Magic: Fire", "Magic: Thunder", "Magic: Water",
+        "\'Ohana Heartbinder", "Dream Heartbinder", "Ocean Heartbinder",
+        "Pixel Heartbinder", "Pride Heartbinder",
+        "Ability: Pole Spin"
```
I believe people will still want Second Chance, Withstand Combo and Block included in important checks.

KiernanBrown replied:
Added Block, Second Chance, and Withstand Combo to the default list of important checks. Went ahead and added Air Slide and Dodge Roll as well since a lot of people have been talking about those defensives/movement options

WaterKH commented on KH3Randomizer/Components/OptionComponents/WorldTreasureSubComponent.razor:
```diff
@@ -9,7 +9,13 @@
+        <td>
+            @Treasure.Id.KeyIdToDisplay()
+            <Tooltip Text="@Treasure.Id.KeyIdToDisplay().GetChestLocation(GetDataTableEnumFromSelection(this.WorldName))">
+                <i class="bi bi-question-circle" style="font-size: 1rem; color: #0366de;"></i>
```
Love that this was gotten around to, I've just been too busy to implement Clara_The_Classy's treasure locations! Thank you!

Commit:
Minor Hint Fixes

Added additional abilities to default important checks.
Fixed some keys for empty categories.
Fixed chest locations that were adjusted in 1.1

Commit:
Update RandomizerService.cs

Removed unnecessary StreamReader lines
